### PR TITLE
Add ingest dbt tests

### LINKED ISSUE #570: Allow users to ingest dbt test results into Soda Cloud via soda-sql
# Context
Users who use dbt have spent time crafting tests in their transformations. Because we don’t want users to repeat themselves to benefit from Soda Cloud’s visibility and incident management we plan to:
- Ingest dbt test results into Soda Cloud.
- Provide clear indications of whether the tests are maintained in soda-sql or ingested from dbt
- Allow collaboration and reporting to be done on those results just as if they were maintained from soda-sql

# User Flow
Users should be able to call a command such as `soda ingest --tool dbt`, provide soda-sql with paths to the dbt artifacts, and have their tests results show up in the Soda Cloud UI. We will have to be clear in our documentation that ingestion should be orchestrated once dbt has been invoked and the transformations have completed. Soda-sql (at least in this first version) will not integrate with the core dbt API. This means it will not orchestrate a dbt run or be called by dbt. Those scenarios may be revisited in further iterations of our integration.

We may also consider creating a new `soda` command such as `soda capture` which would run `soda scan` and follow it with `soda ingest` or a flag on `soda scan` such as `--ingest dbt`. Up for discussion and TBD once minimal isolated approach is working. This is sugar coating at this point.

# Overall Approach
In this v1, we want to keep it simple. dbt generates two valuable artifacts: `manifest.json` and `run_results.json`. When it comes to test results, we can parse the results from the `run_results.json` and use the info in the `manifest.json` to gain knowledge about the table (dataset in Soda's language) that it maps on to.

Once we have parsed all the info we need (see below) we can compose a `TestObject` similar (but not exactly) to the soda-sql `TestObject` that is then exposed to the Soda Cloud backend API. The UI will show the ingested tests on the dataset view, those tests will be clearly identified as `external` and display a dbt logo or whatever UI choice makes sense.

# Specific Requirements
We will want to parse and represent the following information from dbt test results:
- `test_run_at`: timestamp of when the test was run.
- `test_result`: whether the test is passing or failing.
- `severity`: if this info is available (I know it is for freshness but maybe not for all).
- info that allows us to generate a fully qualified table name:
    - `database`
    - `schema`
    - `table`
- `tool_name`: in this case it would be `dbt` (this information will be used by the Soda Cloud UI to display the test provenance and govern available attributes and details)
- `compiled_sql`: dbt should give us access to the compiled sql snippet that it ran for the test. We may want to parse this as we can show it in the Soda Cloud UI and it is valuable information that users can use as a starting point for investigation and issue resolution. If users have enabled `store_failures` we should be able to obtain the `select` statement where they can visualise those failures. If this is possible we could consider adding those to the requirements. More info on this can be found here: https://docs.getdbt.com/docs/building-a-dbt-project/tests#storing-test-failures

# Concious Limitations
- v1 will only support ["generic tests" ](https://docs.getdbt.com/docs/building-a-dbt-project/tests#generic-tests): this means dbt builtins as well as the ones implemented via macros and added to models or columns of those models.
- v1 will not support ["bespoke tests"](https://docs.getdbt.com/docs/building-a-dbt-project/tests#bespoke-tests) unless they are following the best practice of `{{ ref() }}` in their declaration and in that case the mapping should be available in a structured way just like any other tests (`manifest.json` and `run_results.json)
- v1 will not attempt to capture results from `store_failures` (this may be something we offer as a stretch goal as if fits our `failed_rows` model (see last point of [Specific Requirements](#specific-requirements)
- v1 will be limited to ingesting results from dbt-core (OSS). A future iteration will know how to talk to the dbt Cloud API to leverage similar information.

### PR DESCRIPTION
**Pull request by JCZuurmond**
Fixes #570

### CODE REVIEW

File: core/sodasql/cli/cli.py
Comment (JCZuurmond) [inline core/sodasql/cli/cli.py:586]: I do not understand how/why I get non-compiled nodes. I would expect that for tests being executed - and thus part of the run results - that they need to be compiled... I will have a further look at this.

File: core/sodasql/cli/cli.py
Comment (JCZuurmond) [inline core/sodasql/cli/cli.py:588]: What is the test index exactly?

File: core/sodasql/cli/cli.py
Comment (JCZuurmond) [inline core/sodasql/cli/cli.py:590]: We do not have the sql matric name and index

File: core/sodasql/cli/cli.py
Comment (JCZuurmond) [inline core/sodasql/cli/cli.py:593]: I am thinking to change the title in something like: `f"dbt test - {test_run_results.unique_id}"`

File: core/sodasql/cli/cli.py
Comment (JCZuurmond) [inline core/sodasql/cli/cli.py:610]: As value we can add the number of failed rows, not _really_ a value, but makes most sense (for v1)

File: core/sodasql/cli/cli.py
Comment (JCZuurmond) [inline core/sodasql/cli/cli.py:624]: the run results contain time per test, where we have a time per dataset. This also works, but is not _really_ the scan results: when the SQL is executed.

File: core/sodasql/scan/scan_builder.py
Comment (JCZuurmond) [inline core/sodasql/scan/scan_builder.py:30]: I tried to reuse as much as possible from the Scan class. Took out some pieces and put them in separate functions, so that they can be used in different locations more easily

**Review by JCZuurmond — COMMENTED** [review]
I'll move around some stuff and add some more testing. The essence is here though. Let me know if you have any remarks!

File: core/sodasql/cli/cli.py
Comment (JCZuurmond) [reply core/sodasql/cli/cli.py]: This is changed, but it does not have the expected effect: the title in Soda cloud for monitor is not updated accordingly

File: core/sodasql/cli/cli.py
Comment (JCZuurmond) [reply core/sodasql/cli/cli.py:603]: Added this, but it does **not** show up in the Soda cloud.

File: core/sodasql/scan/parser.py
Comment (JCZuurmond) [inline core/sodasql/scan/parser.py:340]: can be removed, not used any more

File: core/sodasql/cli/ingest.py
Comment (JCZuurmond) [inline core/sodasql/cli/ingest.py:174]: move stuff below to separate flush test results function function

File: packages/dbt/sodasql/dbt.py
Comment (JCZuurmond) [inline packages/dbt/sodasql/dbt.py:11]: note to self: double check which nodes `compiled` or `parsed` and other test besides `SchemaTest` we should expect.

File: core/sodasql/cli/ingest.py
Comment (JCZuurmond) [inline core/sodasql/cli/ingest.py:156]: Add more logging for cli user feedback. @vijaykiran : should we add the telemetry stuff here?

File: core/sodasql/cli/ingest.py
Comment (JCZuurmond) [inline core/sodasql/cli/ingest.py:121]: double check if `alias` is correctly used

**Review by JCZuurmond — COMMENTED** [review]
Structure wise the code is getting in a better shape. Most importantly is testing. Two main things that do not work as expected:

1. The title is in Soda cloud is not the same as the title given to the tests.
2. The number of failures is not shown. Maybe because there are no metrics linked?

File: core/sodasql/cli/ingest.py
Comment (bastienboutonnet) [inline core/sodasql/cli/ingest.py:30]: I think this rephrasing might be easier to read:

Map test results to run result

File: core/sodasql/cli/ingest.py
Comment (bastienboutonnet) [reply core/sodasql/cli/ingest.py:156]: I believe we will want to indeed.

**Review by vijaykiran — APPROVED** [review]

File: core/sodasql/cli/ingest.py
Comment (bastienboutonnet) [inline core/sodasql/cli/ingest.py:53]: Because the compiled sql can change from result to result (especially for user defined macro tests), pushing the expression in here will lead to an entirely new test to be created on soda SQL because of how our model works.

We should find another way to generate the ID. As far as I can see, we could just as well leave it as None, or something else.

**Review by bastienboutonnet — COMMENTED** [review]
From a chat with @Antoninj we will need to change the logic of how we create the test id (see comments in the code)

We might also want to make sure we wait for the backend API to be able to accept the source of a test. @Antoninj is working on it.

Finally, I'll add another issue to add support for dbt's `store_failures` but that's out of the initial scope of the PR so I won't be the annoying guy that increases the PR scope at the time that it's almost ready.

File: core/sodasql/cli/ingest.py
Comment (JCZuurmond) [reply core/sodasql/cli/ingest.py:156]: Then I'll leave it up to you or @vijaykiran to do this, I'm ok with someone making those changes to this PR.

File: core/sodasql/cli/ingest.py
Comment (JCZuurmond) [reply core/sodasql/cli/ingest.py:53]: ah, oke, that is good to know. If can confirm that `None` is fine, than that is ok for me. Another option is the raw sql:

```
"raw_sql": "{{ test_accepted_values(**_dbt_schema_test_kwargs) }}{{ config(alias='accepted_values_stg_soda__scan') }}",
```

File: core/sodasql/cli/ingest.py
Comment (JCZuurmond) [reply core/sodasql/cli/ingest.py:121]: According to the [dbt docs](https://docs.getdbt.com/docs/building-a-dbt-project/building-models/using-custom-aliases/), the `alias` is what ultimately defines the table in the warehouse, therefore we should use it.

File: core/sodasql/soda_server_client/soda_server_client.py
Comment (JCZuurmond) [inline core/sodasql/soda_server_client/soda_server_client.py:77]: @vijaykiran : I am not really happy with this change, the schema and database require to have a warehouse object, and some warehouses are not lazily loaded (like parsing connection details or creating a connection). This forces that the ingest has to have a working warehouse set-up, where I think ingest only should have a working Soda cloud connection, besides the dbt artifacts.

File: core/sodasql/soda_server_client/soda_server_client.py
Comment (JCZuurmond) [inline core/sodasql/soda_server_client/soda_server_client.py:65]: I moved these to the arguments, so that they do not require the `warehouse` object

File: packages/dbt/sodasql/dbt.py
Comment (JCZuurmond) [reply packages/dbt/sodasql/dbt.py:11]: For now, I think we are ok, the dbt v1 might cause troubles, see [release notes](https://github.com/dbt-labs/dbt-core/releases/tag/v1.0.0) on tests:

> The two type of test definitions are now "singular" and "generic" (instead of "data" and "schema", respectively). The test_type: selection method accepts test_type:singular and test_type:generic. (It will also accept test_type:schema and test_type:data for backwards compatibility) (#3234, #3880). Not backwards compatible: The --data and --schema flags to dbt test are no longer supported, and tests no longer have the tags 'data' and 'schema' automatically applied.

Comment (JCZuurmond) [reply packages/dbt/sodasql/dbt.py:11]: More general: I asked around in the [dbt slack](https://getdbt.slack.com/archives/C01F91XJ5PY/p1638266195027300) about another possible entrypoint for parsing the artifacts. Conclusion is: that using the node objects is the way to go - for now

File: core/sodasql/cli/cli.py
Comment (JCZuurmond) [reply core/sodasql/cli/cli.py]: I do not have a complete answer for this yet, the `if` statement solves the issue though, so I am approaching this as non-urgent. Moreover, we need to decide what is a proper way to create the test id, as mentioned in one of the comments below, maybe we should remove the `compiled_sql` all together.

**Comment by JCZuurmond** [comment]
Ok, lights are green again, PR is in sync with main. Main issue to resolve: what to do with the test id.

Furthermore, some points that might be covered in another PR:
- add logging to the `ingest` cli function
- add stored failures
- improve integration with dbt cloud, like showing the number of failed rows as measure

File: core/sodasql/cli/ingest.py
Comment (JCZuurmond) [reply core/sodasql/cli/ingest.py:53]: the current `create_test_id` returns something like:

```
        "id": '{"column":"result","test_name":"test.soda.accepted_values_stg_soda__scan__result__pass_fail.81f"}',
```

Is that to be expected?

Comment (bastienboutonnet) [reply core/sodasql/cli/ingest.py]: For the ID, I think the raw SQL may be ok but then again it may change as well. I would prefer to go by the test name and associated column or so.

I'm not sure I understand the latest comment.

Comment (JCZuurmond) [reply core/sodasql/cli/ingest.py]: The last comment shows what the [`create_test_id` method](https://github.com/sodadata/soda-sql/blob/dc31150e3de38060683d27ed1e5bff5b6ec2802e/core/sodasql/scan/parser.py#L357) generates for the test data. It was a fyi, even though more fields are mentioned in [`create_test_id`](https://github.com/sodadata/soda-sql/blob/dc31150e3de38060683d27ed1e5bff5b6ec2802e/core/sodasql/scan/parser.py#L357) for the test data the `column` and `test_name` are the only two that are included.

**Comment by JCZuurmond** [comment]
After discussion with @Antoninj, we decided to use the `unique_id` as test id. The Soda cloud requires a unique identification id for the test, which is the case for the `unique_id` created by dbt.

File: core/sodasql/cli/cli.py
Comment (JCZuurmond) [reply core/sodasql/cli/cli.py]: not relevant anymore.

**Comment by JCZuurmond** [comment]
@vijaykiran : Could you wait with merging, I still have some changes that I would like to make.

**Comment by vijaykiran** [comment]
@JCZuurmond  I marked the PR as draft, when you are ready please switch it  out of draft.

**Review by bastienboutonnet — COMMENTED** [review]
@JCZuurmond are you planning to pass the source (new property now available on backend) in another PR? Personally, I wouldn't want us to start ingesting tests from dbt until they are labelled as coming from dbt so if you do it in another PR I would maybe hold off merging this.

Although, ultimately that's up to @vijaykiran since we are not tagging yet no one is going to use it in production but might be good to make sure we align those.

**Comment by vijaykiran** [comment]
@bastienboutonnet  I'll merge this since I don't want to wait too long. I plan to tag a new release on Tuesday - as long as we can resolve the open issues then we should be good to go.

**Pull request merged**
